File: server/machines/niigata/CurrentStatus.cs
/* Copyright (c) 2020, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
using System;
using System.Linq;
using System.Collections.Generic;
using BlackMaple.MachineFramework;
using System.Collections.Immutable;

namespace BlackMaple.FMSInsight.Niigata
{
  public static class BuildCurrentStatus
  {
    public static CurrentStatus Build(IRepository jobDB, CellState status, FMSSettings settings)
    {
      if (status == null)
      {
        return new CurrentStatus();
      }

      var proc1paths =
        status.UnarchivedJobs.SelectMany(job =>
          Enumerable.Range(1, job.Processes[0].Paths.Count).Select(proc1path => (job, proc1path))
        )
        .OrderBy(x => x.job.RouteStartUTC)
        .ThenBy(x => x.job.Processes[0].Paths[x.proc1path - 1].SimulatedStartingUTC)
        .Select((x, idx) => new { Uniq = x.job.UniqueStr, Proc1Path = x.proc1path, Prec = idx + 1 })
        .ToLookup(x => x.Uniq);

      // jobs
      var jobsByUniq = ImmutableDictionary.CreateBuilder<string, ActiveJob>();
      foreach (var j in status.UnarchivedJobs)
      {
        // completed
        var completed =
          j.Processes.Select(
            proc => new int[proc.Paths.Count] // defaults to fill with zeros
          ).ToArray();
        var evts = jobDB.GetLogForJobUnique(j.UniqueStr);
        foreach (var e in evts)
        {
          if (e.LogType == LogType.LoadUnloadCycle && e.Result == "UNLOAD")
          {
            foreach (var mat in e.Material)
            {
              if (mat.JobUniqueStr == j.UniqueStr)
              {
                var details = jobDB.GetMaterialDetails(mat.MaterialID);
                int matPath = details?.Paths != null && details.Paths.ContainsKey(mat.Process) ? details.Paths[mat.Process] : 1;
                completed[mat.Process - 1][matPath - 1] += 1;
              }
            }
          }
        }

        // precedence
        var proc1precs = proc1paths.Contains(j.UniqueStr) ? proc1paths[j.UniqueStr] : null;
        var precedence =
          j.Processes.Select((proc, procIdx) =>
          {
            if (procIdx == 0)
            {
              return proc.Paths.Select((_, pathIdx) =>
                proc1precs?.FirstOrDefault(x => x.Proc1Path == pathIdx + 1)?.Prec ?? 0L
              ).ToArray();
            }
            else
            {
              return proc.Paths.Select(path =>
                proc1precs?.FirstOrDefault()?.Prec ?? 0L
              ).ToArray();
            }
          }).ToArray();

        // take decremented quantity out of the planned cycles
        int decrQty = j.Decrements?.Sum(d => d.Quantity) ?? 0;
        var newPlanned = j.Cycles - decrQty;

        var started = 0;
        if (status.CyclesStartedOnProc1 != null && status.CyclesStartedOnProc1.TryGetValue(j.UniqueStr, out var startedOnProc1))
        {
          started = startedOnProc1;
        }

        var workorders = jobDB.GetWorkordersForUnique(j.UniqueStr);

        var curJob = j.CloneToDerived<ActiveJob, Job>() with
        {
          CopiedToSystem = true,
          Completed = completed.Select(c => ImmutableList.Create(c)).ToImmutableList(),
          RemainingToStart = decrQty > 0 ? 0 : Math.Max(newPlanned - started, 0),
          Cycles = newPlanned,
          ScheduleId = j.ScheduleId,
          Precedence = precedence.Select(p => ImmutableList.Create(p)).ToImmutableList(),
          Decrements = j.Decrements == null || j.Decrements.Count == 0 ? null : j.Decrements,
          AssignedWorkorders = workorders == null || workorders.Count == 0 ? null : workorders
        };
        jobsByUniq.Add(curJob.UniqueStr, curJob);
      }

      // pallets
      var palletsByName = ImmutableDictionary.CreateBuilder<string, MachineFramework.PalletStatus>();
      foreach (var pal in status.Pallets)
      {
        palletsByName.Add(pal.Status.Master.PalletNum.ToString(), new MachineFramework.PalletStatus()
        {
          Pallet = pal.Status.Master.PalletNum.ToString(),
          FixtureOnPallet = "",
          OnHold = pal.Status.Master.Skip,
          CurrentPalletLocation = pal.Status.CurStation.Location,
          NumFaces = pal.Material.Count > 0 ? pal.Material.Max(m => m.Mat.Location.Face ?? 1) : 0
        });
      }

      return new CurrentStatus()
      {
        TimeOfCurrentStatusUTC = status.Status.TimeOfStatusUTC,
        Jobs = jobsByUniq.ToImmutable(),
        Pallets = palletsByName.ToImmutable(),
        Material =
          status.Pallets.SelectMany(pal => pal.Material)
          .Concat(status.QueuedMaterial)
          .Select(m => m.Mat)
          .Concat(SetLongTool(status))
          .ToImmutableList(),
        QueueSizes = settings.Queues.ToImmutableDictionary(),
        Alarms =
          status.Pallets.Where(pal => pal.Status.Tracking.Alarm).Select(pal => AlarmCodeToString(pal.Status.Master.PalletNum, pal.Status.Tracking.AlarmCode))
          .Concat(
            status.Status.Machines.Values.Where(mc => mc.Alarm)
            .Select(mc => "Machine " + mc.MachineNumber.ToString() + " has an alarm"
            )
          )
          .Concat(
            status.Status.Alarm ? new[] { "ICC has an alarm" } : new string[] { }
          )
          .ToImmutableList()
      };
    }

    private static IEnumerable<InProcessMaterial> SetLongTool(CellState status)
    {
      // tool loads/unloads
      foreach (var pal in status.Pallets.Where(p => p.Status.Master.ForLongToolMaintenance && p.Status.HasWork))
      {
        if (pal.Status.CurrentStep == null) continue;
        switch (pal.Status.CurrentStep)
        {
          case LoadStep load:
            if (pal.Status.Tracking.BeforeCurrentStep)
            {
              yield return new InProcessMaterial()
              {
                MaterialID = -1,
                JobUnique = "",
                PartName = "LongTool",
                Process = 1,
                Path = 1,
                Location = new InProcessMaterialLocation()
                {
                  Type = InProcessMaterialLocation.LocType.Free,
                },
                Action = new InProcessMaterialAction()
                {
                  Type = InProcessMaterialAction.ActionType.Loading,
                  LoadOntoPallet = pal.Status.Master.PalletNum.ToString(),
                  LoadOntoFace = 1,
                  ProcessAfterLoad = 1,
                  PathAfterLoad = 1
                }
              };
            }
            break;
          case UnloadStep unload:
            if (pal.Status.Tracking.BeforeCurrentStep)
            {
              yield return new InProcessMaterial()
              {
                MaterialID = -1,
                JobUnique = "",
                PartName = "LongTool",
                Process = 1,
                Path = 1,
                Location = new InProcessMaterialLocation()
                {
                  Type = InProcessMaterialLocation.LocType.OnPallet,
                  Pallet = pal.Status.Master.PalletNum.ToString(),
                  Face = 1
                },
                Action = new InProcessMaterialAction()
                {
                  Type = InProcessMaterialAction.ActionType.UnloadToCompletedMaterial,
                }
              };
            }
            break;
        }
      }
    }

    private static string AlarmCodeToString(int palletNum, PalletAlarmCode code)
    {
      string msg = " has alarm code " + code.ToString();
      switch (code)
      {
        case PalletAlarmCode.AlarmSetOnScreen:
          msg = " has alarm set by operator";
          break;
        case PalletAlarmCode.M165:
          msg = " has alarm M165";
          break;
        case PalletAlarmCode.RoutingFault:
          msg = " has routing fault";
          break;
        case PalletAlarmCode.LoadingFromAutoLULStation:
          msg = " is loading from auto L/UL station";
          break;
        case PalletAlarmCode.ProgramRequestAlarm:
          msg = " has program request alarm";
          break;
        case PalletAlarmCode.ProgramRespondingAlarm:
          msg = " has program responding alarm";
          break;
        case PalletAlarmCode.ProgramTransferAlarm:
          msg = " has program transfer alarm";
          break;
        case PalletAlarmCode.ProgramTransferFinAlarm:
          msg = " has program transfer alarm";
          break;
        case PalletAlarmCode.MachineAutoOff:
          msg = " at machine with auto off";
          break;
        case PalletAlarmCode.MachinePowerOff:
          msg = " at machine which is powered off";
          break;
        case PalletAlarmCode.IccExited:
          msg = " can't communicate with ICC";
          break;
      }
      return "Pallet " + palletNum.ToString() + msg;
    }
  }
}

File: client/insight/src/cell-status/current-status.ts
/* Copyright (c) 2021, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import { JobsBackend } from "../network/backend";
import {
  InProcessMaterial,
  ICurrentStatus,
  IInProcessMaterial,
  ILogEntry,
  LogType,
  LocType,
  ActiveJob,
} from "../network/api";
import { atom, DefaultValue, RecoilValueReadOnly, selectorFamily, TransactionInterface_UNSTABLE } from "recoil";
import { last30JobComment } from "./scheduled-jobs";
import { conduit } from "../util/recoil-util";
import type { ServerEventAndTime } from "./loading";

const currentStatusRW = atom<Readonly<ICurrentStatus>>({
  key: "current-status",
  default: {
    timeOfCurrentStatusUTC: new Date(),
    jobs: {},
    pallets: {},
    material: [],
    alarms: [],
    queues: {},
  },
});
export const currentStatus: RecoilValueReadOnly<ICurrentStatus> = currentStatusRW;

export const currentStatusJobComment = selectorFamily<string | null, string>({
  key: "current-status-job-comment",
  get:
    (uniq) =>
    ({ get }) =>
      get(currentStatus).jobs[uniq]?.comment ?? null,
  set:
    (uniq) =>
    ({ set }, newVal) => {
      const newComment = newVal instanceof DefaultValue || newVal === null ? "" : newVal;

      set(currentStatusRW, (st) => {
        const oldJob = st.jobs[uniq];
        if (oldJob) {
          const newJob = new ActiveJob(oldJob);
          newJob.comment = newComment;
          return { ...st, jobs: { ...st.jobs, [uniq]: newJob } };
        } else {
          return st;
        }
      });

      set(last30JobComment(uniq), newComment);

      void JobsBackend.setJobComment(uniq, newComment);
    },
  cachePolicy_UNSTABLE: { eviction: "lru", maxSize: 1 },
});

export const setCurrentStatus = conduit<Readonly<ICurrentStatus>>((t: TransactionInterface_UNSTABLE, st) =>
  t.set(currentStatusRW, st)
);

export const updateCurrentStatus = conduit<ServerEventAndTime>(
  (t: TransactionInterface_UNSTABLE, { evt }: ServerEventAndTime) => {
    if (evt.logEntry) {
      t.set(currentStatusRW, processEventsIntoCurrentStatus(evt.logEntry));
    } else if (evt.newCurrentStatus) {
      t.set(currentStatusRW, evt.newCurrentStatus);
    }
  }
);

function processEventsIntoCurrentStatus(
  entry: Readonly<ILogEntry>
): (curSt: Readonly<ICurrentStatus>) => Readonly<ICurrentStatus> {
  return (curSt) => {
    const mats = new Map<number, Readonly<InProcessMaterial>>();
    function adjustMat(id: number, f: (mat: Readonly<IInProcessMaterial>) => Readonly<IInProcessMaterial>) {
      if (mats.size === 0) {
        for (const m of curSt.material) {
          mats.set(m.materialID, m);
        }
      }
      const oldMat = mats.get(id);
      if (oldMat !== undefined) {
        mats.set(id, new InProcessMaterial(f(oldMat)));
      }
    }

    switch (entry.type) {
      case LogType.PartMark:
        for (const m of entry.material) {
          adjustMat(m.id, (inmat) => ({ ...inmat, serial: entry.result }));
        }
        break;

      case LogType.OrderAssignment:
        for (const m of entry.material) {
          adjustMat(m.id, (inmat) => ({ ...inmat, workorderId: entry.result }));
        }
        break;

      case LogType.Inspection:
      case LogType.InspectionForce:
        if (entry.result.toLowerCase() === "true" || entry.result === "1") {
          let inspType: string;
          if (entry.type === LogType.InspectionForce) {
            inspType = entry.program;
          } else {
            inspType = (entry.details || {}).InspectionType;
          }
          if (inspType) {
            for (const m of entry.material) {
              adjustMat(m.id, (inmat) => ({
                ...inmat,
                signaledInspections: [...inmat.signaledInspections, inspType],
              }));
            }
          }
        }
        break;
    }

    if (mats.size === 0) {
      return curSt;
    } else {
      return {
        ...curSt,
        material: Array.from(mats.values()),
      };
    }
  };
}

export interface QueueReordering {
  readonly queue: string;
  readonly matId: number;
  readonly newIdx: number;
}

export const reorderQueuedMatInCurrentStatus = conduit<QueueReordering>(
  (t: TransactionInterface_UNSTABLE, { queue, matId, newIdx }: QueueReordering) => {
    t.set(currentStatusRW, (curSt) => {
      const oldMat = curSt.material.find((i) => i.materialID === matId);
      if (!oldMat || oldMat.location.type !== LocType.InQueue) {
        return curSt;
      }
      if (oldMat.location.currentQueue === queue && oldMat.location.queuePosition === newIdx) {
        return curSt;
      }

      const oldQueue = oldMat.location.currentQueue;
      const oldIdx = oldMat.location.queuePosition;
      if (oldIdx === undefined || oldQueue === undefined) {
        return curSt;
      }

      const newMats = curSt.material.map((m) => {
        if (
          m.location.type !== LocType.InQueue ||
          m.location.queuePosition === undefined ||
          m.location.currentQueue === undefined
        ) {
          return m;
        }

        if (m.materialID === matId) {
          return new InProcessMaterial({
            ...m,
            location: { type: LocType.InQueue, currentQueue: queue, queuePosition: newIdx },
          } as IInProcessMaterial);
        }

        let idx = m.location.queuePosition;
        // old queue material is moved down
        if (m.location.currentQueue === oldQueue && m.location.queuePosition > oldIdx) {
          idx -= 1;
        }
        // new queue material is moved up to make room
        if (m.location.currentQueue === queue && idx >= newIdx) {
          idx += 1;
        }

        if (idx !== m.location.queuePosition) {
          return new InProcessMaterial({
            ...m,
            location: { ...m.location, queuePosition: idx },
          } as IInProcessMaterial);
        } else {
          return m;
        }
      });

      return { ...curSt, material: newMats };
    });
  }
);
